<a href="https://colab.research.google.com/gist/ievgenya20/39f03fc3b71f7cc5e417fc916beccb44/final_homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
import numpy as np
import pandas as pd

Считываем данные с drive.google по ссылке

In [32]:
google_url = 'https://drive.google.com/file/d/1INgo03nal-vwFJe7Lec5vOUtOwfJdUr1/view?usp=sharing'
url = 'https://drive.google.com/uc?id=' + google_url.split('/')[-2]
df = pd.read_csv(url)
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

Удаляем дубликаты строк

In [33]:
df1 = df.drop_duplicates().reset_index(drop=True)
df1.head(5)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [34]:
df1[df1.duplicated() == True]

Empty DataFrame
Columns: [satisfaction_level, last_evaluation, number_project, average_montly_hours, time_spend_company, Work_accident, left, promotion_last_5years, department, salary]
Index: []

In [35]:
df1.isnull().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
department               0
salary                   0
dtype: int64

Проверяем на пустые столбцы

In [49]:
df1.dropna(axis=1).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11991 entries, 0 to 11990
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     11991 non-null  float64
 1   last_evaluation        11991 non-null  float64
 2   number_project         11991 non-null  int64  
 3   average_montly_hours   11991 non-null  int64  
 4   time_spend_company     11991 non-null  int64  
 5   Work_accident          11991 non-null  int64  
 6   left                   11991 non-null  int64  
 7   promotion_last_5years  11991 non-null  int64  
 8   department             11991 non-null  object 
 9   salary                 11991 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 936.9+ KB


Статистические данные массива

In [36]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11991 entries, 0 to 11990
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     11991 non-null  float64
 1   last_evaluation        11991 non-null  float64
 2   number_project         11991 non-null  int64  
 3   average_montly_hours   11991 non-null  int64  
 4   time_spend_company     11991 non-null  int64  
 5   Work_accident          11991 non-null  int64  
 6   left                   11991 non-null  int64  
 7   promotion_last_5years  11991 non-null  int64  
 8   department             11991 non-null  object 
 9   salary                 11991 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 936.9+ KB


Отсекаем текстовые и столбцы с бинарными (0/1) данными

In [37]:
df1.iloc[:,:5].describe().round(2).loc[['50%', 'mean', 'min', 'max', 'std']]

satisfaction_level  last_evaluation  number_project  \
50%                 0.66             0.72            4.00   
mean                0.63             0.72            3.80   
min                 0.09             0.36            2.00   
max                 1.00             1.00            7.00   
std                 0.24             0.17            1.16   

      average_montly_hours  time_spend_company  
50%                 200.00                3.00  
mean                200.47                3.36  
min                  96.00                2.00  
max                 310.00               10.00  
std                  48.73                1.33

Рассчитайте и визуализировать корреляционную матрицу для
количественных переменных

In [38]:
df1.iloc[:,:5].corr().round(3)

satisfaction_level  last_evaluation  number_project  \
satisfaction_level                 1.000            0.095          -0.133   
last_evaluation                    0.095            1.000           0.270   
number_project                    -0.133            0.270           1.000   
average_montly_hours              -0.006            0.265           0.332   
time_spend_company                -0.153            0.097           0.189   

                      average_montly_hours  time_spend_company  
satisfaction_level                  -0.006              -0.153  
last_evaluation                      0.265               0.097  
number_project                       0.332               0.189  
average_montly_hours                 1.000               0.103  
time_spend_company                   0.103               1.000

Визуализируем результаты

In [39]:
corr = df1[['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company']].corr()
corr.style.background_gradient(cmap='coolwarm')

Наибольший уровень корреляции:

average_monthly_hours и number_project - 0.332

Наименьший уровень уровень корреляции:

time_spend_company и satisfaction_level - (-0.133)

Рассчитайте сколько сотрудников работает в каждом
департаменте.

In [40]:
df1.groupby('department').size()

department
IT              976
RandD           694
accounting      621
hr              601
management      436
marketing       673
product_mng     686
sales          3239
support        1821
technical      2244
dtype: int64

Показать распределение сотрудников по зарплатам.

In [41]:
df1.salary.value_counts()

low       5740
medium    5261
high       990
Name: salary, dtype: int64

Показать распределение сотрудников по зарплатам в каждом
департаменте по отдельности. Строим сводную таблицу по пересечению департаменту и зарплат.

In [42]:
df.pivot_table(index='department', columns='salary', aggfunc='size')

salary       high   low  medium
department                     
IT             83   609     535
RandD          51   364     372
accounting     74   358     335
hr             45   335     359
management    225   180     225
marketing      80   402     376
product_mng    68   451     383
sales         269  2099    1772
support       141  1146     942
technical     201  1372    1147

Проверить гипотезу, что сотрудники с высоким окладом
проводят на работе больше времени, чем сотрудники с низким
окладом

Среднее количество рабочих часов сотрудников c низким окладом

In [44]:
df1[df1.salary == 'low']['average_montly_hours'].mean()

200.4391986062718

Среднее количество рабочих часов сотрудников c высоким окладом

In [46]:
df1[df1.salary == 'high']['average_montly_hours'].mean()

200.06767676767677

За нулевую гипотезу принимаем, что уровень оклада не оказывает влияния на среднее количество часов, проведенных на работе

Также задаем две переменные salary_low и salary_high

In [47]:
salary_low = df1[df1['salary'] == 'low']['average_montly_hours']
salary_high = df1[df1['salary'] == 'high']['average_montly_hours']

Через ttest проводим проверку гипотезы под pvalue

In [52]:
import numpy as np
import scipy.stats as stats

In [54]:
result = stats.ttest_ind(salary_low, salary_high, equal_var=False)

a = 0.05

if (result.pvalue < a):
    print('Отвергаем гипотезу H0. В среднем проведённое на работе время отличается в зависимости от уровня зарплаты.')
else:
    print('Принимаем гипотезу H0. В среднем проведённое на работе время одинаково для всех работников.')

Принимаем гипотезу H0. В среднем проведённое на работе время одинаково для всех работников.


Рассчитать следующие показатели среди уволившихся и не
уволившихся сотрудников (по отдельности):

*   Доля сотрудников с повышением за последние 5 лет
*   Средняя степень удовлетворенности
*   Среднее количество проектов





Два массива данных: уволившиеся (left=1), работающие (left=0)

In [56]:
actual = df[(df['left'] == 0)]
fired =  df[(df['left'] == 1)]

Формируем сводную таблицу из 3-х строк по заданным условиям.
Столбца 2: уволенные и действующие

In [58]:
massive = pd.DataFrame()
rows = []

rows.append(
    {
    'Показатель': 'Доля получивших повышение за 5 лет',
    'Действующие': round(actual.promotion_last_5years.value_counts(normalize=True)[1], 3),
    'Уволившиеся': round(fired.promotion_last_5years.value_counts(normalize=True)[1], 3)
    }
)

rows.append(
    {
    'Показатель': 'Средняя степень удовлетворенности',
    'Действующие': round(actual.satisfaction_level.mean(), 3),
    'Уволившиеся': round(fired.satisfaction_level.mean(), 3)
    }
)

rows.append(
    {
    'Показатель': 'Среднее количество проектов',
    'Действующие': round(actual.number_project.mean(), 3),
    'Уволившиеся': round(fired.number_project.mean(), 3)
    }
)


for row in rows:
    massive = pd.concat([massive, pd.DataFrame([row])])

massive

Показатель  Действующие  Уволившиеся
0  Доля получивших повышение за 5 лет        0.026        0.005
0   Средняя степень удовлетворенности        0.667        0.440
0         Среднее количество проектов        3.787        3.856

Разделить данные на тестовую и обучающую выборки
Построить модель LDA, предсказывающую уволился ли
сотрудник на основе имеющихся факторов (кроме department и
salary)
Оценить качество модели на тестовой выборки

In [59]:
X = df1.iloc[:,[0,1,2,3,4,5,7]]
y = df1['left']

In [63]:
from sklearn.model_selection import train_test_split
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [68]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [69]:
model = LinearDiscriminantAnalysis()
model.fit(X_train, y_train)

LinearDiscriminantAnalysis()

In [70]:
model.predict(X_test)

array([1, 0, 0, ..., 0, 1, 0])

In [71]:
model.score(X_test, y_test).round(2)

0.83

In [77]:
from sklearn.metrics import accuracy_score

In [78]:
accuracy_score(y_test, model.predict(X_test)).round(2)

0.83